brush-core/src/builtins/ulimit.rs

```diff
@@ -0,0 +1,206 @@
+            Unit::Number => format!("(-{})", self.short),
+        };
+        let resource = self.get(hard)?;
+        println!("{:<26}{:>16} {}", self.description, unit, resource);
+        Ok(())
+    }
+}
+
+const ALL_RESOURCES: [ResourceDescription; 4] = [
+    ResourceDescription::CORE,
+    ResourceDescription::FSIZE,
+    ResourceDescription::NOFILE,
+    ResourceDescription::RSS,
+];
+
+#[derive(Debug, Clone, Copy)]
+enum LimitValue {
+    Unset,
+    Unlimited,
+    Soft,
+    Hard,
+    Value(u64),
+}
+
+impl FromStr for LimitValue {
+    type Err = <u64 as FromStr>::Err;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let v = match s {
+            "" => LimitValue::Unset,
+            "unlimited" => LimitValue::Unlimited,
+            "soft" => LimitValue::Soft,
+            "hard" => LimitValue::Hard,
+            _ => LimitValue::Value(s.parse()?),
+        };
+        Ok(v)
+    }
+}
+
+/// Modify shell resource limits.
+///
+/// Provides control over the resources available to the shell and processes
+/// it creates, on systems that allow such control.
+#[derive(Parser, Debug)]
+pub(crate) struct ULimitCommand {
+    /// use the `soft` resource limit
+    #[arg(short = 'S')]
+    soft: bool,
+    /// use the `hard` resource limit
+    #[arg(short = 'H')]
+    hard: bool,
+    /// all current limits are reported
+    #[arg(short)]
+    all: bool,
+    /// the maximum size of core files created
+    #[arg(short, default_missing_value = "", num_args(0..=1))]
+    core: Option<LimitValue>,
+    /// the maximum size of files written by the shell and its children
+    #[arg(short = 'f', default_missing_value = "", num_args(0..=1))]
+    file_size: Option<LimitValue>,
+    /// the maximum resident set size
+    #[arg(short = 'm', default_missing_value = "", num_args(0..=1))]
+    rss: Option<LimitValue>,
+    /// the maximum number of open file descriptors
+    #[arg(short = 'n', default_missing_value = "", num_args(0..=1))]
+    file_open: Option<LimitValue>,
+
+    /// argument for the implicit limit (`-f`)
+    limit: Option<LimitValue>,
+}
+
+impl builtins::Command for ULimitCommand {
+    async fn execute(
+        &self,
+        _context: commands::ExecutionContext<'_>,
+    ) -> Result<crate::builtins::ExitCode, crate::error::Error> {
+        let exit_code = builtins::ExitCode::Success;
+        let mut resources_to_set = Vec::new();
+        let mut resources_to_get = Vec::new();
+
+        let mut set_or_get = |val, descr| match val {
+            Some(LimitValue::Unset) => resources_to_get.push(descr),
+            Some(v) => resources_to_set.push((descr, v)),
+            None => {}
+        };
+
+        set_or_get(self.core, ResourceDescription::CORE);
+        set_or_get(self.file_size, ResourceDescription::FSIZE);
+        set_or_get(self.rss, ResourceDescription::RSS);
+        set_or_get(self.file_open, ResourceDescription::NOFILE);
+
+        if self.all {
```

Comment:
If we folded checking `self.all` into the body of `set_or_get`, that could allow us to remove `ALL_RESOURCES` and leave only one place where we need to explicitly list all resource types (i.e., on lines 169-172). It would also complicate `set_or_get`.

What do you think about that trade-off?

Reply:
It is simple enough